### DISCUSSION TITLE
Incompatible React versions: The "react" and "react-dom" packages must have the exact same version.

### DISCUSSION THREAD

**Discussion by dnlwllms** [webpack/webpack#19078] [Q&A]
Webpack raised the following error:

> Incompatible React versions: The "react" and "react-dom" packages must have the exact same version. Instead got:
> - react: 19.0.0-rc-7283a213-20241206
> - react-dom: 19.0.0
> Learn more: https://react.dev/warnings/version-mismatch

Next.js is failing to import the React module built with Webpack.
This is an issue that occurs after upgrading to React 19 Stable version.

Directly import the module my local, it works normally, but this happens when I access the bundle from Webpack build.

I would appreciate it if you could share the solution.

**Comment by ramonmulia** [comment]
@dnlwllms Did you get this solved?
I'm facing the same issue :|

**Comment by YoussefKababe** [comment]
Facing this too...

**Comment by khanakia** [comment]
I am facing this issue with Next.js

![Screenshot 2025-01-05 at 11 29 19 PM](https://github.com/user-attachments/assets/88b4aa7b-1277-41ae-8ae8-0e5d53fe660c)

**Reply by YoussefKababe** [reply]
@khanakia Are you by any chance importing code from your own React library? Is Fasto your own library or something like that? I was able to solve my issue, but in my case it was a problem with the build config of my own React components library, some code from some React related package that implements this version check was leaking into my build bundle.

**Reply by khanakia** [reply]
Yes, It's my library. 

But i use Next.js where I import my UI library. I upgraded to the Next.js 15 which uses the `19.0.0-rc-65e06cb7-20241218` React

1. My UI library has this in package.json

```json
  "react": "^19.0.0",
  "react-dom": ">=19.0.0",
```

2. Check the Next.js Code here
https://github.com/vercel/next.js/blob/4cbaaa118d2138edd32263da0dc4b2ecb9ddbc2b/package.json#L271C1-L272C1

**Reply by YoussefKababe** [reply]
Then I might have the solution for you! I'm building my library using Vite/Rollup, so here's what I did to solve the issue:

```javascript
  rollupOptions: {
      external: (id) => {
        return (
          id === 'react' ||
          id === 'react-dom' ||
          id.startsWith('react/') ||
          id.startsWith('react-dom/')
      },
      output: {
        globals: {
          react: 'React',
          'react-dom': 'ReactDOM',
          'react/jsx-runtime': 'jsxRuntime',
        },
      },
    },
  },
```

This will externalize all React related packages and exclude their code from your build bundle. The NextJS app where you're importing your library should provide these packages already.

I hope this helps!

**Reply by khanakia** [reply]
i had already react them set as Externals however i used your code instead of mine

Now i get this `Attempted import error: 'createRoot' is not exported from 'react-dom' (imported as 'ReactDOM').`


![Screenshot 2025-01-06 at 10 51 23 PM](https://github.com/user-attachments/assets/ed40f6df-b9fc-4307-9353-f240d9aa41ef)

**Reply by khanakia** [reply]
It's because of this patch code i added too https://ant.design/docs/react/v5-for-19/

**Reply by YoussefKababe** [reply]
Looks like you just need to import `createRoot` from `react-dom/client` instead of `react-dom` in your library!

**Reply by khanakia** [reply]
I am not importing react-dom anywhere except for that patch.tsx file I created as suggested by antd

![Screenshot 2025-01-07 at 11 24 51 AM](https://github.com/user-attachments/assets/969e36bc-b017-4991-a438-144a25928375)

**Reply by khanakia** [reply]
Ok, i have fixed it thanks

**Reply by itzfeltrin** [reply]
Thanks @YoussefKababe !!! It worked for me

**Comment by muhuanlegend** [comment]
i deleted the package-lock.json file and deleted node modules ..then made sure in my package.json file react and react dom have the same version and i removed the ^ character then did a fresh install npm i and that solved my headache 
<img width="1352" height="713" alt="image" src="https://github.com/user-attachments/assets/c9baa484-e31c-4944-ab72-dc94cb88c763" />
